### PR TITLE
Support for OMOP ES batches

### LINKED ISSUE #231: Accept OMOP ES batches
### Definition of Done / Acceptance Criteria

- [ ] PIXL CLI reads from multiple batches, within a single OMOP ES. Populating the queues as if it were a single batch
- [ ] PIXL CLI copies the public files from multiple batches into the extracts directory 
- [ ] Assuming #192 is completed, then should refactor this to copy the directory structure rather thsn globbing

### Testing

- [ ] Will need to have a test example with multiple batches
- [ ] If we have tests for copying parquet files via SFTP then this should also be tested, but if not this can be done when we implement the parquet uploading

### Documentation

Update CLI documentation

### Details and Comments

Expected structure for a single project which is split up over 3 batches (this would happen when there is a large amount of data and we want to ensure ease of transferring. 
```
└── omop_extract
    ├── extract_1
    │   ├── extract_summary.json
    │   ├── private
    │   └── public
    ├── extract_2
    │   ├── extract_summary.json
    │   ├── private
    │   └── public
    └── extract_3
        ├── extract_summary.json
        ├── private
        └── public
```
Where `omop_extract` is the directory that is passed into `pixl populate`. `extract_*` are expected to be literal dir names.

### PR DESCRIPTION
**Pull request by jeremyestein**
Will fix #231 

TODO:
- [x] Split test OMOP extracts into batches, delete stray extra copy
- [x] Document the different bits of test data (batches) and why we have them
- [x] Implement the multi batch detection and reading/merging
- [ ] Tests....
  - [ ] increase test coverage a bit for the public methods in _io.py?
- [ ] Clarify batch dir naming (`extract_N` vs `batch_N`), change code and docs when decided
- [x] Check assumption that timestamp will be identical for all batches within an extract

### CODE REVIEW

File: cli/src/pixl_cli/_io.py [hunk 1]
```diff
@@ -47,20 +47,75 @@ def messages_from_state_file(filepath: Path) -> list[Message]:
...
+    :raises RuntimeError: if config files are non-existent or contradictory
+    """
+    log_filename = "extract_summary.json"
+    single_batch_logfile = extract_path / log_filename
+    if single_batch_logfile.exists():
+        project_name, omop_es_timestamp = _config_from_log_file(single_batch_logfile)
+        return project_name, omop_es_timestamp, [extract_path]
+
+    # should it really be 'extract_*'?
+    batch_dirs = list(extract_path.glob("batch_*"))
```
Comment (stefpiatek) [inline cli/src/pixl_cli/_io.py:70]: Yeah we're given `extract_*`, though we could ask if OMOP ES could change it? Worth dropping a message in pixl-omop channel in UCLH Foundry slack.

File: cli/src/pixl_cli/_io.py [hunk 2]
```diff
@@ -47,20 +47,75 @@ def messages_from_state_file(filepath: Path) -> list[Message]:
...
+        err = f"No batched or unbatched log files found in {extract_path}"
+        raise RuntimeError(err)
+
+    distinct_extract_ids = set(extract_ids)
+    if len(distinct_extract_ids) != 1:
+        err = (
+            f"Found {len(batch_dirs)} log files with different IDs: "
+            f"Batch dirs: {batch_dirs}, IDs: {extract_ids}"
+        )
+        raise RuntimeError(err)
```
Comment (stefpiatek) [inline cli/src/pixl_cli/_io.py:83]: Nice

File: cli/src/pixl_cli/_io.py [hunk 3]
```diff
@@ -47,20 +47,75 @@ def messages_from_state_file(filepath: Path) -> list[Message]:
     return [deserialise(line) for line in filepath.open().readlines() if string_is_non_empty(line)]
 
 
-def config_from_log_file(parquet_path: Path) -> tuple[str, datetime]:
-    log_file = parquet_path / "extract_summary.json"
+def determine_batch_structure(extract_path: Path) -> tuple[str, datetime, list[Path]]:
```
Comment (stefpiatek) [inline cli/src/pixl_cli/_io.py:50]: maybe `parse_batch_structure`?

File: cli/src/pixl_cli/_io.py (hunk 1 above)
Comment (stefpiatek) [reply cli/src/pixl_cli/_io.py:70]: Also, should we sort these?

File: cli/src/pixl_cli/_io.py [hunk 4]
```diff
@@ -47,20 +47,75 @@ def messages_from_state_file(filepath: Path) -> list[Message]:
...
+
+    # should it really be 'extract_*'?
+    batch_dirs = list(extract_path.glob("batch_*"))
+    extract_ids = [_config_from_log_file(log_file / log_filename) for log_file in batch_dirs]
+    # There should be one or more log files, all with the same identifiers
+    if not extract_ids:
+        err = f"No batched or unbatched log files found in {extract_path}"
+        raise RuntimeError(err)
+
+    distinct_extract_ids = set(extract_ids)
```
Comment (stefpiatek) [inline cli/src/pixl_cli/_io.py:77]: might be nice to log how many directories are found?

File: cli/tests/test_messages_from_parquet.py [hunk 5]
```diff
@@ -15,30 +15,48 @@
...
+    ("batches", "single_batch", "expected_message_indexes"),
+    [
+        (["batch_1", "batch_2"], False, [0, 1]),
+        (["batch_1"], True, [0]),
+        (["batch_1"], False, [0]),
+        (["batch_2"], True, [1]),
+        (["batch_2"], False, [1]),
+    ],
+)
+def test_messages_from_parquet(
```
Comment (stefpiatek) [inline cli/tests/test_messages_from_parquet.py:39]: chef's kiss

**Review by stefpiatek — APPROVED** [review]
Looking good, will need to either change to using `export_*` or ensure that OMOP ES changes how they do their batches before merging

**Review by milanmlft — APPROVED** [review]
Thanks for the detailed docstrings! ❤️ 
Just one small suggestion; maybe add the example file structure from the issue to the `cli/README.md` as well? And clarify that we support this multi-batch structure.

```
└── omop_extract
    ├── extract_1
    │   ├── extract_summary.json
    │   ├── private
    │   └── public
    ├── extract_2
    │   ├── extract_summary.json
    │   ├── private
    │   └── public
    └── extract_3
        ├── extract_summary.json
        ├── private
        └── public
```

File: cli/src/pixl_cli/_io.py (hunk 4 above)
Comment (jeremyestein) [reply cli/src/pixl_cli/_io.py:77]: + which directories are currently being exported.
+ Add a test case where a batch is dodgy (eg. parquet truncated), make sure no DSH upload happens, and that trtrying works.

**Pull request closed without merging**